README.md

```diff
@@ -1,20 +1,124 @@
 
-  2. Ensure phoenix_gen_socket_client is started before your application:
+def project do
+  [
+    deps: [
+      {:phoenix_gen_socket_client, github: "aircloak/phoenix_gen_socket_client"}
+      # ...
+    ],
+    application: [
+      applications: [:phoenix_gen_socket_client, # ...],
+      # ...
+    ]
+  ]
+end
+```
 
-        def application do
-          [applications: [:phoenix_gen_socket_client]]
-        end
+You also need to add the transport library (for example a websocket client) as a dependency. Out of the box, the adapter for [sanmiguel/websocket_client](https://github.com/sanmiguel/websocket_client) is provided, but you still need to include the library yourself:
 
+```elixir
+def project do
+  [
+    deps: [
+      {:websocket_client, github: "sanmiguel/websocket_client", tag: "1.1.0"},
+      # ...
+    ],
+    application: [
+      applications: [:websocket_client, # ...],
+      # ...
+    ]
+  ]
+end
+```
+
+
+## Usage
+
+This library is designed for flexibility, so the usage is a bit more involved. In general, you need to:
+
+1. Implement a callback module for the `Phoenix.Channels.GenSocketClient` behaviour.
+2. Start the socket process somewhere in your supervision tree.
+3. Interact with the server from callback functions which are running in the socket process.
+
+A simple demo is available in the [example](example) folder. Here, we'll present some general ideas. For more details, refer to the documentation. It is currently not available online, but you can clone this repo and build it locally with `MIX_ENV=docs mix docs`.
+
+__Note__: In the subsequent code snippets we assume that `Phoenix.Channels.GenSocketClient` is aliased, so we use `GenSocketClient`.
+
+
+### Starting the socket process
+
+Starting the process works similar to other behaviours, such as `GenServer`:
+
+```elixir
+GenSocketClient.start_link(
+  callback_module,
+  transport_module,
+  arbitrary_argument,
+  socket_opts, # defaults to []
+  gen_server_opts # defaults to []
+)
+```
+
+For `transport_module` you can pass any module which implements the `Phoenix.Channels.GenSocketClient.Transport` behaviour. Out of the box, you have the module `Phoenix.Channels.GenSocketClient.Transport.WebsocketClient` available.
+
+The code above will start another process where the `init/1` function of the `callback_module` is invoked. This function needs to provide the initial state and the socket url.
+
+The socket url must also include the transport suffix. For example, if in the server socket you have declared socket with `socket "/my_socket", ...`, then the url for the websocket transport would be `ws://server_url/my_socket/websocket`.
+
+
+### Connection life-cycle
+
+After `init/1` returns, the transport process is not immediately created. You need to do this manually from one of `handle_*` callbacks. To initially establish the connection, you can send yourself a message from `init/1`, and then request the connection to be established in `handle_info/2`:
```

Comment:
We don't want to start the connection on `init/1` since this means the supervisor starting the socket process will be blocked until we connect. So on a laggy network that might take awhile.

We could add some smartness of our own where on `{:connect, ...}` the behaviour sends itself a message and then handles it explicitly. I was worried this might be a bit magical, and there might be some edge cases lurking. At least with this approach it's quite clear what goes one. 

Perhaps it would suffice if I explain the reasoning better?